activesupport/lib/active_support/current_attributes.rb

```diff
@@ -107,39 +109,7 @@ def attribute(*names)
           raise ArgumentError, "Restricted attribute names: #{invalid_attribute_names.join(", ")}"
         end
 
-        ActiveSupport::CodeGenerator.batch(generated_attribute_methods, __FILE__, __LINE__) do |owner|
-          names.each do |name|
-            owner.define_cached_method(name, namespace: :current_attributes) do |batch|
-              batch <<
-                "def #{name}" <<
-                "attributes[:#{name}]" <<
-                "end"
-            end
-            owner.define_cached_method("#{name}=", namespace: :current_attributes) do |batch|
-              batch <<
-                "def #{name}=(value)" <<
-                "attributes[:#{name}] = value" <<
-                "end"
-            end
-          end
-        end
-
-        ActiveSupport::CodeGenerator.batch(singleton_class, __FILE__, __LINE__) do |owner|
-          names.each do |name|
-            owner.define_cached_method(name, namespace: :current_attributes_delegation) do |batch|
-              batch <<
-                "def #{name}" <<
-                "instance.#{name}" <<
-                "end"
-            end
-            owner.define_cached_method("#{name}=", namespace: :current_attributes_delegation) do |batch|
-              batch <<
-                "def #{name}=(value)" <<
-                "instance.#{name} = value" <<
-                "end"
-            end
-          end
-        end
+        names.each { |name| define_attribute_methods name }
```

Comment:
@byroot could you review these changes, since you're in this block's `git blame`? I'm curious about how to re-implement the delegation method definitions. At the moment, the tests pass, but rely on the first `#method_missing` to handle the delegation.

Reply:
> At the moment, the tests pass, but rely on the first `#method_missing` to handle the delegation.

YOu mean from the class to the instance? Yeah define method (or eval) inside `method_missing` is a terrible pattern I wish to eliminate. 

I'll set a reminder to look at your PR tomorrow.

Reply:
Ah, but we were eagerly defining the delegators for the attributes, it's only for unexpected method that we lazy defined them (which is less bad but not great).

I think you can simply call `delegate` right after the line we're commenting, and you can use the private `as:` argument to get optimized delegation, should produce about the same result.

Reply:
@byroot I think I understand your suggestion in principle, but I'm struggling to craft  the statement.

I'm trying this, but it's failing:

```ruby
delegate name, "#{name}=", to: :instance, as: singleton_class
# => activesupport/lib/active_support/core_ext/module/delegation.rb:220:in `public_instance_method': undefined method `world' for class `#<Class:CurrentAttributesTest::Current>' (NameError)
# 
#             receiver_class.public_instance_method(method)
#                           ^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
> I'm trying this, but it's failing:

`as:` take the class of the object that is supposed to receive the call, so here it would have been `as: self`.

Reply:
Actually if you still to want to reduce duplication in that module, you can do the `delegate` part only, that would do it.

Reply:
@byroot I've opened https://github.com/rails/rails/pull/50685 to cherry-pick that change.

Reply:
> Yeah define method (or eval) inside `method_missing` is a terrible pattern I wish to eliminate.

@byroot I've explored an alternative to method missing in https://github.com/rails/rails/pull/50686.